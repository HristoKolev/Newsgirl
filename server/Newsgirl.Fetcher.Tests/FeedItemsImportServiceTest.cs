namespace Newsgirl.Fetcher.Tests;

using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Shared;
using Xdxd.DotNet.Testing;
using Xunit;

public class FeedItemsImportServiceTest : AppDatabaseTest
{
    [Fact]
    public async Task GetFeedsForUpdate_Returns_All_Feeds()
    {
        var importService = new FeedItemsImportService(this.Db);

        var feeds = Enumerable.Range(1, 10).Select(i => new FeedPoco
        {
            FeedItemsHash = i,
            FeedContentHash = i,
            FeedName = $"feed {i}",
            FeedUrl = $"url {i}",
        }).ToList();

        await this.Db.BulkInsert(feeds);

        var resultFeeds = await importService.GetFeedsForUpdate();

        Snapshot.Match(resultFeeds);
    }

    [Fact]
    public async Task ImportItems_Copies_Items_Correctly()
    {
        var importService = new FeedItemsImportService(this.Db);

        var feeds = Enumerable.Range(1, 10).Select(i => new FeedPoco
        {
            FeedName = $"feed {i}",
            FeedUrl = $"url {i}",
        }).ToList();

        // Insert one by one in order to set the PK value.
        foreach (var feed in feeds)
        {
            await this.Db.Insert(feed);
        }

        var updates = new List<FeedUpdateModel>
        {
            new FeedUpdateModel
            {
                NewItems = Enumerable.Range(1, 10).Select(i => new FeedItemPoco
                {
                    FeedID = 1,
                    FeedItemDescription = $"desc {i}",
                    FeedItemStringID = $"string id {i}",
                    FeedItemStringIDHash = i,
                    FeedItemTitle = $"title {i}",
                    FeedItemUrl = $"url {i}",
                    FeedItemAddedTime = StubHelper.Date3000,
                }).ToList(),
                Feed = feeds.First(x => x.FeedID == 1),
                NewFeedItemsHash = 1,
                NewFeedContentHash = 1,
            },
            new FeedUpdateModel
            {
                NewItems = Enumerable.Range(100, 10).Select(i => new FeedItemPoco
                {
                    FeedID = 2,
                    FeedItemDescription = $"desc {100 + i}",
                    FeedItemStringID = $"string id {i}",
                    FeedItemStringIDHash = i,
                    FeedItemTitle = $"title {100 + i}",
                    FeedItemUrl = $"url {100 + i}",
                    FeedItemAddedTime = StubHelper.Date3000,
                }).ToList(),
                Feed = feeds.First(x => x.FeedID == 2),
                NewFeedItemsHash = 2,
                NewFeedContentHash = 2,
            },
            new FeedUpdateModel
            {
                NewItems = Enumerable.Range(200, 10).Select(i => new FeedItemPoco
                {
                    FeedID = 3,
                    FeedItemDescription = null,
                    FeedItemStringID = $"string id {i}",
                    FeedItemStringIDHash = i,
                    FeedItemTitle = $"title {i}",
                    FeedItemUrl = null,
                    FeedItemAddedTime = StubHelper.Date3000,
                }).ToList(),
                Feed = feeds.First(x => x.FeedID == 3),
                NewFeedItemsHash = 3,
                NewFeedContentHash = 3,
            },
        };

        foreach (var update in updates)
        {
            await importService.ApplyUpdate(update);
        }

        var resultFeeds = this.Db.Poco.Feeds.OrderByDescending(x => x.FeedID).ToList();
        var resultFeedItems = this.Db.Poco.FeedItems.OrderByDescending(x => x.FeedItemID).ToList();

        Snapshot.Match(new { resultFeeds, resultFeedItems });
    }

    [Fact]
    public async Task GetMissingFeedItems_Returns_Correct_Result()
    {
        var importService = new FeedItemsImportService(this.Db);

        var feeds = Enumerable.Range(1, 10).Select(i => new FeedPoco
        {
            FeedName = $"feed {i}",
            FeedUrl = $"url {i}",
        }).ToList();

        await this.Db.BulkInsert(feeds);

        var feedItems = Enumerable.Range(1, 10).Select(i => new FeedItemPoco
        {
            FeedID = 1,
            FeedItemStringID = $"string id {i}",
            FeedItemStringIDHash = i,
            FeedItemDescription = $"desc {i}",
            FeedItemTitle = $"title {i}",
            FeedItemUrl = $"url {i}",
            FeedItemAddedTime = StubHelper.Date3000,
        }).ToList();

        feedItems.AddRange(Enumerable.Range(30, 10).Select(i => new FeedItemPoco
        {
            FeedID = 2,
            FeedItemStringID = $"string id {i}",
            FeedItemStringIDHash = i,
            FeedItemDescription = $"desc {i}",
            FeedItemTitle = $"title {i}",
            FeedItemUrl = $"url {i}",
            FeedItemAddedTime = StubHelper.Date3000,
        }));

        await this.Db.BulkInsert(feedItems);

        long[] hashes =
        {
            1, 2, 3, 31, 32, 33,
        };

        var missingFeedItems = await importService.GetMissingFeedItems(1, hashes);

        Array.Sort(missingFeedItems);

        Snapshot.Match(missingFeedItems);
    }
}