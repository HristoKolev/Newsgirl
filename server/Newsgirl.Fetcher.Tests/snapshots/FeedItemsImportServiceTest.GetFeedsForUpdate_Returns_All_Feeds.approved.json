[
  {
    "FeedContentHash": 10,
    "FeedID": 10,
    "FeedItemsHash": 10,
    "FeedName": "feed 10",
    "FeedUrl": "url 10"
  },
  {
    "FeedContentHash": 9,
    "FeedID": 9,
    "FeedItemsHash": 9,
    "FeedName": "feed 9",
    "FeedUrl": "url 9"
  },
  {
    "FeedContentHash": 8,
    "FeedID": 8,
    "FeedItemsHash": 8,
    "FeedName": "feed 8",
    "FeedUrl": "url 8"
  },
  {
    "FeedContentHash": 7,
    "FeedID": 7,
    "FeedItemsHash": 7,
    "FeedName": "feed 7",
    "FeedUrl": "url 7"
  },
  {
    "FeedContentHash": 6,
    "FeedID": 6,
    "FeedItemsHash": 6,
    "FeedName": "feed 6",
    "FeedUrl": "url 6"
  },
  {
    "FeedContentHash": 5,
    "FeedID": 5,
    "FeedItemsHash": 5,
    "FeedName": "feed 5",
    "FeedUrl": "url 5"
  },
  {
    "FeedContentHash": 4,
    "FeedID": 4,
    "FeedItemsHash": 4,
    "FeedName": "feed 4",
    "FeedUrl": "url 4"
  },
  {
    "FeedContentHash": 3,
    "FeedID": 3,
    "FeedItemsHash": 3,
    "FeedName": "feed 3",
    "FeedUrl": "url 3"
  },
  {
    "FeedContentHash": 2,
    "FeedID": 2,
    "FeedItemsHash": 2,
    "FeedName": "feed 2",
    "FeedUrl": "url 2"
  },
  {
    "FeedContentHash": 1,
    "FeedID": 1,
    "FeedItemsHash": 1,
    "FeedName": "feed 1",
    "FeedUrl": "url 1"
  }
]